code/modules/mob/living/carbon/xenomorph/castes/defiler/abilities_defiler.dm

```diff
@@ -155,9 +163,86 @@
 
 	succeed_activate()
 	add_cooldown()
-	
+
 	new /obj/effect/alien/egg/gas(A.loc)
 	qdel(alien_egg)
 
 	GLOB.round_statistics.defiler_inject_egg_neurogas++
 	SSblackbox.record_feedback("tally", "round_statistics", 1, "defiler_inject_egg_neurogas")
+
+// ***************************************
+// *********** Reagent selection
+// ***************************************
+/datum/action/xeno_action/select_reagent
+	name = "Select Reagent"
+	action_icon_state = "select_reagent0"
+	mechanics_text = "Selects which reagent to inject. Hemodile slows by 25%, increased to 50% with neurotoxin present, and deals 20% of dmage received as stamina damage. Transvitox after ~4s converts brute/burn damage to toxin based on 40% of damage received up to 45 toxin on target."
+	use_state_flags = XACT_USE_BUSY
+
+/datum/action/xeno_action/select_reagent/update_button_icon()
+	var/mob/living/carbon/xenomorph/X = owner
+	var/atom/A = X.selected_reagent
+	button.overlays.Cut()
+	button.overlays += image('icons/mob/actions.dmi', button, initial(A.name))
+	return ..()
+
+/datum/action/xeno_action/select_reagent/action_activate()
+	var/mob/living/carbon/xenomorph/X = owner
+	var/list/available_reagents = X.xeno_caste.available_reagents_define
+	var/i = available_reagents.Find(X.selected_reagent)
+	if(length(available_reagents) == i)
+		X.selected_reagent = available_reagents[1]
+	else
```

Comment:
This code is confusing and doesnt work properly

Reply:
Can you elaborate more on how it is not working? It is switching reagents as intended on my end.

Reply:
var/i = available_reagents.Find(X.selected_reagent)
	if(length(available_reagents) == i)
		X.selected_reagent = available_reagents[1]

find the selected reagent and set it to i
if the length of available reagents == reference to the selected reagent
select the first reagent in the list

This should be length == 1, i should be removed or renamed (?)